File: src/test_hcp.py
import os
import sys
import argparse

# Automatically add the current directory (src) to the Python path
sys.path.insert(0, os.path.abspath(os.path.dirname(__file__)))

from libs.encodings.cardinality_one.bimander import BimanderEncoder
from libs.encodings.cardinality_one.binomial import BinomialEncoder
from libs.encodings.cardinality_one.binary import BinaryEncoder
from libs.encodings.cardinality_one.commander import CommanderEncoder
from libs.encodings.cardinality_one.product import ProductEncoder
from libs.encodings.cardinality_one.sequential import SequentialEncoder
from libs.encodings.cardinality_one.hybrid import HybridEncoder
from libs.encodings.cardinality_one.scl import NSCEncoding
from libs.encodings.cardinality_one.pblib import PbLibEncoder
from libs.encodings.cardinality_one.heule import HeuleEncoder

from libs.models.successor.base import SuccessorMethod
from libs.models.successor.unary import Unary
from libs.models.successor.lfsr import LFSR
from libs.models.successor.binary_adder import BinaryAdder
from libs.models.successor.binary_adder_original import BinaryAdderOriginal
from libs.models.successor.crt import CRT
from libs.models.successor.vee import Vee
from libs.models.successor.inc_crt import IncCRT
from libs.models.successor.inc_ocrt import IncOCRT

from libs.utils.common import get_files_from_folder
from solver import HcpSolver, IncHcpSolver


def main():
    parser = argparse.ArgumentParser(description="Solve Hamiltonian Cycle Problem (HCP) using SAT Encodings.")
    parser.add_argument("-s", "--successor", choices=["unary", "lfsr", "binary_adder", "binary_adder_original", "crt", "vee", "inc_crt", "inc_ocrt"], default="binary_adder", help="Successor method to use (default: binary_adder)")
    parser.add_argument("-e", "--encoder", choices=["bimander", "binomial", "binary", "commander", "product", "sequential", "hybrid", "pblib", "scl", "heule"], default="bimander", help="Polymorphic encoder to use (default: bimander)")
    parser.add_argument("-d", "--data", type=str, default="src/data/fhcpcs", help="Folder containing graph instances (default: src/data/fhcpcs)")
    parser.add_argument("--solver", choices=["glucose", "cadical"], default="glucose", help="Solver backend (default: glucose)")
    parser.add_argument("-l", "--log", type=str, default="log-file.txt", help="Path to log file (default: log-file.txt)")

    args = parser.parse_args()
    successor_type = args.successor
    encoder_type = args.encoder
    data_folder = args.data
    solver_backend = args.solver
    log_file = args.log

    # 1. Instantiate Card-One Encoder
    if encoder_type == "bimander":
        encoder = BimanderEncoder()
    elif encoder_type == "binomial":
        encoder = BinomialEncoder()
    elif encoder_type == "binary":
        encoder = BinaryEncoder()
    elif encoder_type == "commander":
        encoder = CommanderEncoder()
    elif encoder_type == "product":
        encoder = ProductEncoder()
    elif encoder_type == "sequential":
        encoder = SequentialEncoder()
    elif encoder_type == "hybrid":
        encoder = HybridEncoder()
    elif encoder_type == "pblib":
        encoder = PbLibEncoder()
    elif encoder_type == "scl":
        encoder = NSCEncoding()
    elif encoder_type == "heule":
        encoder = HeuleEncoder()
    else:
        raise ValueError(f"Unknown encoder type: {encoder_type}")

    # 2. Get Files
    if not os.path.exists(data_folder):
        print(f"Data folder or file {data_folder} not found.")
        sys.exit(1)

    if os.path.isfile(data_folder):
        listFiles = [data_folder]
    else:
        listFiles = get_files_from_folder(data_folder)
    if not listFiles:
        print(f"No files found in {data_folder}.")
        sys.exit(0)

    print(f"Found {len(listFiles)} graph instances in {data_folder}.")
    print(f"Using Successor: {successor_type.upper()} | Encoder: {encoder_type.upper()} | Solver: {solver_backend.upper()}")

    for path in listFiles:  # Process all files
        print("\n" + "=" * 50)
        print("Start build clauses for:", path.split("/")[-1])

        # 3. Instantiate Successor Method
        if successor_type == "unary":
            successor = Unary()
        elif successor_type == "lfsr":
            successor = LFSR()
        elif successor_type == "crt":
            successor = CRT()
        elif successor_type == "inc_crt":
            successor = IncCRT()
        elif successor_type == "inc_ocrt":
            successor = IncOCRT(vee_encoder=NSCEncoding(), crt_encoder=encoder)
        elif successor_type == "binary_adder_original":
            successor = BinaryAdderOriginal()
        elif successor_type == "vee":
            successor = Vee()
        else:
            successor = BinaryAdder()

        # 4. Set up Solver
        if getattr(successor, 'is_incremental', False):
            hcpSolver = IncHcpSolver(successor, encoder)
        else:
            hcpSolver = HcpSolver(successor, encoder)
        hcpSolver.graph.load_graph_from_file(path)

        # 5. Solve
        if solver_backend == "glucose":
            result = hcpSolver.solve()
        else:
            result = hcpSolver.solve_cadical()

        # 6. Print Results
        print("Solving finished with status:", result["status"])
        print(f"  - Variables: {result['nofVariables']}")
        print(f"  - Clauses: {result['nofClauses']}")
        if result["time"] is not None:
            print(f"  - Solve Time: {result['time']}s")

        if result["status"] == "SAT":
            hcpSolver.print_result(result["model"], hcpSolver.successor.getH, result)
            status = "valid" if result['vOfHC'] == hcpSolver.graph.v else "invalid"
            print(f"Result verification: {status.upper()}")

        # 7. Save to log file
        if log_file:
            with open(log_file, "a") as f:
                filename = path.split("/")[-1]
                status_str = "valid" if result.get('vOfHC') == hcpSolver.graph.v else "None"
                f.write(f"{filename.ljust(30)} {str(result['nofVariables']).ljust(10)} {str(result['nofClauses']).ljust(10)} {result['status'].ljust(10)}  {str(result['time']).ljust(10)} {status_str.ljust(10)}\n")


if __name__ == "__main__":
    main()

File: src_cpp/test_hcp.cpp
#include "libs/encodings/cardinality_one/BimanderEncoder.hpp"
#include "libs/encodings/cardinality_one/BinomialEncoder.hpp"
#include "libs/encodings/cardinality_one/BinaryEncoder.hpp"
#include "libs/encodings/cardinality_one/CommanderEncoder.hpp"
#include "libs/encodings/cardinality_one/ProductEncoder.hpp"
#include "libs/encodings/cardinality_one/SequentialEncoder.hpp"
#include "libs/encodings/cardinality_one/HybridEncoder.hpp"
#include "libs/encodings/cardinality_one/NSCEncoding.hpp"
#include "libs/encodings/cardinality_one/PbLibEncoder.hpp"
#include "libs/encodings/cardinality_one/HeuleEncoder.hpp"

#include "libs/models/successor/Unary.hpp"
#include "libs/models/successor/LFSR.hpp"
#include "libs/models/successor/BinaryAdder.hpp"
#include "libs/models/successor/BinaryAdderOriginal.hpp"
#include "libs/models/successor/CRT.hpp"
#include "libs/models/successor/Vee.hpp"
#include "libs/models/successor/IncCRT.hpp"
#include "libs/models/successor/OCRT.hpp"
#include "libs/models/successor/IncOCRT.hpp"
#include "libs/models/successor/OUnary.hpp"
#include "libs/models/successor/PruningCRT.hpp"
#include "libs/models/successor/IncPruningCRT.hpp"

#include "libs/utils/Common.hpp"
#include "solver.hpp"

#include <iostream>
#include <memory>
#include <chrono>
#include <iomanip>
#include <fstream>
#include <algorithm>

std::unique_ptr<CardinalityOneEncoder> make_encoder(const std::string& type) {
    if (type == "bimander") return std::make_unique<BimanderEncoder>();
    if (type == "binomial") return std::make_unique<BinomialEncoder>();
    if (type == "binary") return std::make_unique<BinaryEncoder>();
    if (type == "commander") return std::make_unique<CommanderEncoder>();
    if (type == "product") return std::make_unique<ProductEncoder>();
    if (type == "sequential") return std::make_unique<SequentialEncoder>();
    if (type == "hybrid") return std::make_unique<HybridEncoder>();
    if (type == "pblib") return std::make_unique<PbLibEncoder>();
    if (type == "scl") return std::make_unique<NSCEncoding>();
    if (type == "heule") return std::make_unique<HeuleEncoder>();
    throw std::runtime_error("Unknown encoder: " + type);
}

std::unique_ptr<SuccessorMethod> make_successor(const std::string& type, CardinalityOneEncoder* encoder, int cycle_len) {
    if (type == "unary") return std::make_unique<Unary>(encoder);
    if (type == "lfsr") return std::make_unique<LFSR>(encoder);
    if (type == "crt") return std::make_unique<CRT>(encoder, cycle_len);
    if (type == "inc_crt") return std::make_unique<IncCRT>(encoder, cycle_len);
    if (type == "vee") return std::make_unique<Vee>(encoder);
    if (type == "ocrt") return std::make_unique<OCRT>(nullptr, encoder, cycle_len);
    if (type == "inc_ocrt") return std::make_unique<IncOCRT>(nullptr, encoder, cycle_len);
    if (type == "ounary") return std::make_unique<OUnary>(nullptr);
    if (type == "pruning_crt") return std::make_unique<PruningCRT>(encoder, cycle_len);
    if (type == "inc_pruning_crt") return std::make_unique<IncPruningCRT>(encoder, cycle_len);
    if (type == "binary_adder_original") return std::make_unique<BinaryAdderOriginal>(encoder);
    if (type == "binary_adder") return std::make_unique<BinaryAdder>(encoder);
    throw std::runtime_error("Unknown successor: " + type);
}

void print_help(char* argv[]) {
    std::cout << "Usage: " << argv[0] << " [options]" << std::endl;
    std::cout << "Options:" << std::endl;
    std::cout << "  -s, --successor <type>   Successor method: unary, lfsr, binary_adder, binary_adder_original, crt, vee, ocrt, inc_crt, inc_ocrt, pruning_crt, ounary (default: binary_adder)" << std::endl;
    std::cout << "  -e, --encoder <type>     Cardinality-one encoder: bimander, binomial, binary, commander, product, sequential, hybrid, pblib, scl, heule (default: bimander)" << std::endl;
    std::cout << "  -d, --data <path>        Folder containing graph instances or single file (default: src/data/fhcpcs)" << std::endl;
    std::cout << "  -l, --log <path>         Path to log file (default: log-file.txt)" << std::endl;
    std::cout << "  -c, --cycle <limit>      CRT cycle limit override" << std::endl;
    std::cout << "  -t, --timeout <seconds>  Solver timeout limit in seconds (default: 600)" << std::endl;
    std::cout << "  --solver-mode <0|1>      0 = Propagator (default), 1 = CEGAR (only for incremental solvers)" << std::endl;
    std::cout << "  --opt-level <0|1|2>      Optimization level (default: 1)" << std::endl;
    std::cout << "  -h, --help               Show this help message" << std::endl;
}

int main(int argc, char* argv[]) {
    std::string successor_type = "binary_adder";
    std::string encoder_type = "bimander";
    std::string data_path = "src/data/fhcpcs";
    std::string log_file = "log-file.txt";
    int cycle_len = -1;
    int solver_mode = 0; // Propagator
    int timeout_seconds = 600;
    int opt_level = 1;
    int seed = 0;
    int assume_u = 0;
    int assume_v = 0;

    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "-s" || arg == "--successor") {
            if (i + 1 < argc) successor_type = argv[++i];
        } else if (arg == "-e" || arg == "--encoder") {
            if (i + 1 < argc) encoder_type = argv[++i];
        } else if (arg == "-d" || arg == "--data") {
            if (i + 1 < argc) data_path = argv[++i];
        } else if (arg == "-l" || arg == "--log") {
            if (i + 1 < argc) log_file = argv[++i];
        } else if (arg == "-c" || arg == "--cycle") {
            if (i + 1 < argc) cycle_len = std::stoi(argv[++i]);
        } else if (arg == "-t" || arg == "--timeout") {
            if (i + 1 < argc) timeout_seconds = std::stoi(argv[++i]);
        } else if (arg == "--solver-mode") {
            if (i + 1 < argc) solver_mode = std::stoi(argv[++i]);
        } else if (arg == "--opt-level") {
            if (i + 1 < argc) opt_level = std::stoi(argv[++i]);
        } else if (arg == "--seed") {
            if (i + 1 < argc) seed = std::stoi(argv[++i]);
        } else if (arg == "--assume") {
            if (i + 1 < argc) {
                std::string assume_str = argv[++i];
                size_t comma_pos = assume_str.find(',');
                if (comma_pos != std::string::npos) {
                    assume_u = std::stoi(assume_str.substr(0, comma_pos));
                    assume_v = std::stoi(assume_str.substr(comma_pos + 1));
                }
            }
        } else if (arg == "-h" || arg == "--help") {
            print_help(argv);
            return 0;
        }
    }

    auto encoder = make_encoder(encoder_type);
    auto listFiles = get_files_from_folder(data_path);

    if (listFiles.empty()) {
        std::cout << "No graph instances found in: " << data_path << std::endl;
        return 0;
    }

    std::cout << "Found " << listFiles.size() << " graph instances in " << data_path << "." << std::endl;
    std::cout << "Using Successor: " << successor_type << " | Encoder: " << encoder_type << " | Opt-level: " << opt_level << std::endl;

    for (const auto& path : listFiles) {
        std::cout << "\n==================================================" << std::endl;
        std::cout << "Start build clauses for: " << fs::path(path).filename().string() << std::endl;

        int local_cycle = cycle_len;
        if (local_cycle == -1 && (successor_type == "inc_crt" || successor_type == "inc_ocrt" || successor_type == "inc_pruning_crt")) {
            local_cycle = 12;
        }

        auto successor = make_successor(successor_type, encoder.get(), local_cycle);
        
        Graph graph;
        graph.load_graph_from_file(path);
        
        int orig_v = graph.v;
        if (opt_level >= 1) {
            graph.preprocess();
            if (graph.v < orig_v) {
                std::cout << "Preprocessing (SOL-5): reduced vertices from " << orig_v << " to " << graph.v << std::endl;
            }
        }

        SolverResult result;
        if (successor->is_incremental) {
            IncHcpSolver solver(successor.get(), encoder.get(), opt_level);

            result = solver.solve_incremental(graph, solver_mode, timeout_seconds, seed, assume_u, assume_v);
            std::cout << "Solving finished with status: " << result.status << std::endl;
            std::cout << "  - Variables: " << result.nofVariables << std::endl;
            std::cout << "  - Clauses: " << result.nofClauses << std::endl;
            std::cout << "  - Solve Time: " << result.time << "s" << std::endl;
            if (result.status == "SAT") {
                solver.print_result(graph, result);
            }
        } else {
            HcpSolver solver(successor.get(), encoder.get(), opt_level);
            result = solver.solve(graph, timeout_seconds, seed, assume_u, assume_v);
            std::cout << "Solving finished with status: " << result.status << std::endl;
            std::cout << "  - Variables: " << result.nofVariables << std::endl;
            std::cout << "  - Clauses: " << result.nofClauses << std::endl;
            std::cout << "  - Solve Time: " << result.time << "s" << std::endl;
            if (result.status == "SAT") {
                solver.print_result(graph, result);
            }
        }

        if (!log_file.empty()) {
            std::ofstream log(log_file, std::ios::app);
            if (log.is_open()) {
                std::string filename = fs::path(path).filename().string();
                log << std::left << std::setw(30) << filename
                    << std::setw(10) << result.nofVariables
                    << std::setw(10) << result.nofClauses
                    << std::setw(10) << result.status
                    << std::setw(10) << result.time
                    << std::endl;
            }
        }
    }

    return 0;
}
